bench: refactor random number generation in `stats/base/dists/triangular`

G4URAV001 opened the pull request:
… in JS benchmarks

Resolves #4989 .

## Description

> What is the purpose of this pull request?

This pull request:

-   This pull request improves random number generation in JS benchmarks for `stats/base/dists/triangular`. Specifically:

- Moves random number generation out of benchmarking loops to prevent interference with benchmark results.
- Uses `@stdlib/random/base/uniform` and `@stdlib/random/base/discrete-uniform` instead of `randu` expressions.
- Ensures generated random values maintain the same range as existing values for consistency.


## Related Issues

> Does this pull request have any related issues?

>This pull request:

- Resolves #4989 
- References #4837 and #4955 as examples of similar improvements.


## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

   
   ✅ Build Native Add-on.
   ✅ Run JavaScript Benchmarks.
   ✅ Run JavaScript Native Benchmarks.


* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

G4URAV001 committed:
bench(stats/base/dists/triangular): refactor random number generation in JS benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

G4URAV001 commented:
i have added the following which [stdlib-bot](https://github.com/stdlib-bot) pointed out 
"-   [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)"

Neerajpathak07 commented:
> i have added the following which [stdlib-bot](https://github.com/stdlib-bot) pointed out "- [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)"

Hey, Could you pls add a space after the hyphen Ig that's what's causing the error.

G4URAV001 commented:
> > i have added the following which [stdlib-bot](https://github.com/stdlib-bot) pointed out "- [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)"

 >>hey, Could you pls add a space after the hyphen Ig that's what's causing the error.

I think it was for the main PR. i didn't had there initially, i added it now. there

Neerajpathak07 commented:
Apparently the CI errors is specific to the current PR and not affected by other ones. I suppose the change should be made here.

G4URAV001 commented:
> Apparently the CI errors is specific to the current PR and not affected by other ones. I suppose the change should be made h

ya i think i made it now in current PR. It should work now

G4URAV001 commented:
@anandkaranubc  Can you review the PR now ?

anandkaranubc commented:
> @anandkaranubc Can you review the PR now ?

On it!

anandkaranubc commented:
@G4URAV001 Can you also update the title to:

bench: refactor random number generation in JS benchmarks for `stats/base/dists/triangular`

which is just adding backticks around stats/base/dists/triangular

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/benchmark/benchmark.js:
```diff
@@ -34,17 +35,26 @@ bench( pkg, function benchmark( b ) {
-		max = min + ( randu() * 40.0 ) + EPS;
-		mode = min + ( ( max - min ) * randu() );
-		y = cdf( x, min, max, mode );
+		y = cdf( x[ i%len ], min[ i%len ], max[ i%len ], mode[ i%len ] );
```
```suggestion
		y = cdf( x[ i % len ], min[ i % len ], max[ i % len ], mode[ i % len ] );
```

G4URAV001 replied:
if I do this change then it will give eslint error of maximum allowed length of the line. Also in some benchmark where it was giving the same issue it was already without the spaces, that's why i did that to rest of the places.
Should I still do these changes

Neerajpathak07 replied:
add this flag after a space at the end of the line to disable the eslint check.
```
// eslint-disable max-len
```

anandkaranubc replied:
You can leave those lines as they are, as done in:

https://github.com/stdlib-js/stdlib/blob/8514fc990e8358e7e7b32f704ad8b8edc376ae98/lib/node_modules/%40stdlib/stats/base/dists/triangular/logpdf/benchmark/benchmark.js#L58

The goal here is to improve readability and consistency. This spacing is (I believe) to ensure consistency across different namespaces. However, in cases like these (where you face this lint error), you can group those elements together, as shown in the example above.

This [discussion](https://github.com/stdlib-js/stdlib/pull/5140#discussion_r1959243271) will certainly be helpful.

@kgryte, please correct me if I'm wrong, but writing `x[ i%len ]` is "better" than `x[ i % len ]`. However, the latter is followed to maintain consistency across different packages, right?

kgryte replied:
@anandkaranubc Yes, at least in `stats/base/dists`, we've been doing `x[ i % len ]`. Either is fine.

@Neerajpathak07 The comment should be `// eslint-disable-line max-len`. What you suggested disables the lint rule for the entire file, which is likely NOT what we want.

anandkaranubc replied:
Thanks!

Neerajpathak07 replied:
ahh my bad!!

G4URAV001 replied:
Thanks everyone

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/ctor/benchmark/benchmark.js:
```diff
@@ -95,17 +105,21 @@ bench( pkg+'::set:a', function benchmark( bm ) {
 	}
 	bm.toc();
-	if ( isnan( y ) ) {
+	if ( isnan( y[ i % len ] ) ) {
```
```suggestion
	if ( isnan( y ) ) {
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/ctor/benchmark/benchmark.js:
```diff
@@ -152,17 +167,21 @@ bench( pkg+'::set:b', function benchmark( bm ) {
+	len = 100;
+	y = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		y[ i ] = uniform( c + EPS, c + 100.0);
```
```suggestion
		y[ i ] = uniform( c + EPS, c + 100.0 );
```

There should be empty spaces at the beginning and end of brackets.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/ctor/benchmark/benchmark.js:
```diff
@@ -152,17 +167,21 @@ bench( pkg+'::set:b', function benchmark( bm ) {
 	}
 	bm.toc();
-	if ( isnan( y ) ) {
+	if ( isnan( y[ i % len ] ) ) {
```
```suggestion
	if ( isnan( y ) ) {
```

This checks whether the random number generation was successful.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/ctor/benchmark/benchmark.js:
```diff
@@ -209,17 +229,21 @@ bench( pkg+'::set:c', function benchmark( bm ) {
 	}
 	bm.toc();
-	if ( isnan( y ) ) {
+	if ( isnan( y[ i % len ] ) ) {
```
Same comment and applies throughout the PR.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/benchmark/benchmark.js:
```diff
@@ -34,17 +35,26 @@ bench( pkg, function benchmark( b ) {
-		max = min + ( randu() * 40.0 ) + EPS;
-		mode = min + ( ( max - min ) * randu() );
-		y = logcdf( x, min, max, mode );
+		y = logcdf( x[ i%len ], min[ i%len ], max[ i%len ], mode[ i%len ] );
```
```suggestion
		y = logcdf( x[ i % len ], min[ i % len ], max[ i % len ], mode[ i % len ] );
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/benchmark/benchmark.js:
```diff
@@ -47,10 +47,10 @@ bench( pkg, function benchmark( b ) {
-		min[ i ] = randu() * 10.0;
-		max[ i ] = min[ i ] + ( randu() * 40.0 ) + EPS;
-		mode[ i ] = min[ i ] + ( ( max[ i ] - min[ i ] ) * randu() );
+		t[ i ] = uniform( 0.0, 5.0);
```
```suggestion
		t[ i ] = uniform( 0.0, 5.0 );
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/benchmark/benchmark.js:
```diff
@@ -81,11 +82,15 @@ bench( pkg+':factory', function benchmark( b ) {
+	len = 100;
+	t = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		t[ i ] = uniform( 0.0, 5.0);
```
```suggestion
		t[ i ] = uniform( 0.0, 5.0 );
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/benchmark/benchmark.native.js:
```diff
@@ -56,10 +56,10 @@ bench( pkg+'::native', opts, function benchmark( b ) {
-		min[ i ] = randu() * 10.0;
-		max[ i ] = min[ i ] + ( randu() * 40.0 ) + EPS;
-		mode[ i ] = min[ i ] + ( ( max[ i ] - min[ i ] ) * randu() );
+		t[ i ] = uniform( 0.0, 5.0);
```
Same comment as above

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/pdf/benchmark/benchmark.js:
```diff
@@ -47,10 +47,10 @@ bench( pkg, function benchmark( b ) {
 	mode = new Float64Array( len );
 
 	for ( i = 0; i < len; i++ ) {
```
```suggestion
	mode = new Float64Array( len );
	for ( i = 0; i < len; i++ ) {
```

There should be no empty lines between initializations. This applies throughout the PR.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/benchmark/benchmark.native.js:
```diff
@@ -56,10 +56,10 @@ bench( pkg+'::native', opts, function benchmark( b ) {
 	mode = new Float64Array( len );
 
 	for ( i = 0; i < len; i++ ) {
```
```suggestion
	mode = new Float64Array( len );
	for ( i = 0; i < len; i++ ) {
```

Same comment as above and applies elsewhere

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/quantile/benchmark/benchmark.js:
```diff
@@ -34,17 +35,27 @@ bench( pkg, function benchmark( b ) {
+	len = 100;
+	p = new Float64Array( len );
+	min = new Float64Array( len );
+	max = new Float64Array( len );
+	mode = new Float64Array( len );
+
+	for ( i = 0; i < len; i++ ) {
+		p[ i ] = uniform( 0.0, 1.0 );
+		min[ i ] = uniform( 0.0, 10.0 );
+		max[ i ] = uniform( min[ i ] + EPS, min[ i ] + 40.0 );
+		mode[ i ] = uniform( min[ i ], max[ i ] );
+	}
+
 	b.tic();
 	for ( i = 0; i < b.iterations; i++ ) {
-		p = randu();
-		min = randu() * 10.0;
-		max = min + ( randu() * 40.0 ) + EPS;
-		mode = min + ( ( max - min ) * randu() );
-		y = quantile( p, min, max, mode );
+		y = quantile( p[ i%len ][ i%len ], max[ i%len ], mode[ i%len ] );
```
Spacing needs to be fixed here.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/quantile/benchmark/benchmark.js:
```diff
@@ -70,11 +82,15 @@ bench( pkg+':factory', function benchmark( b ) {
 	for ( i = 0; i < b.iterations; i++ ) {
-		p = randu();
-		y = myquantile( p );
+		y = myquantile( p[ i % len] );
```
```suggestion
		y = myquantile( p[ i % len ] );
```

G4URAV001 committed:
fixup! Made the suggested changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

G4URAV001 commented:
@anandkaranubc Did the changes, please review this

anandkaranubc commented:
> @anandkaranubc Did the changes, please review this

On it!

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/skewness/benchmark/benchmark.js:
```diff
@@ -43,10 +43,11 @@ bench( pkg, function benchmark( b ) {
 	min = new Float64Array( len );
 	max = new Float64Array( len );
 	mode = new Float64Array( len );
+
 	for ( i = 0; i < len; i++ ) {
```
```suggestion
	mode = new Float64Array( len );
	for ( i = 0; i < len; i++ ) {
```

anandkaranubc reviewed (changes requested):
One small change. Rest, everything looks good! Thanks!

G4URAV001 committed:
fixup! made the suggested change

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Neerajpathak07 commented:
Not something related to this PR But,
I think we should achieve 100% code coverage for the following packages `ctor`,`median` and `mgf`.
@anandkaranubc @kgryte what do you guys think!!

kgryte commented:
@Neerajpathak07 Agreed. Feel free to submit separate PRs for addressing any gaps in coverage.

G4URAV001 commented:
@anandkaranubc hey please review it I think I made all the corrections

G4URAV001 committed:
Merge branch 'develop' into G4URAV001/issue4989

Signed-off-by: Gaurav <gaurav70380@gmail.com>

G4URAV001 commented:
Had some bugs in last commit
